### PR TITLE
Bump Gradle version, migrate to Kotlin DSL, refactor publishing

### LINKED ISSUE #566: Clean up core artifacts
The goal of this task is to revise and clean up gradle tasks responsible for building and publishing artifacts.
The most important part of it will be creation of standalone `api` artifact that should be the only dependency needed for plugin development.

### PR DESCRIPTION
**Pull request by kamilok1965**
Fixes #566 and #584

### CODE REVIEW

**Comment by Kordyjan** [comment]
I think it is good place to remove `.idea` folder.

**Comment by kamilok1965** [comment]
Agreed, however it was decided not to do it some time ago

File: runners/maven-plugin/build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    dependsOn(tasks.compileKotlin, tasks.compileJava)
+    from("$buildDir/classes/kotlin", "$buildDir/classes/java")
+    into("$mavenBuildDir/classes/java")
+
+    preserve {
+        include("**/*.class")
+    }
+}
+
+val helpMojo by tasks.registering(org.jetbrains.CrossPlatformExec::class) {
```
Comment (Kordyjan) [inline runners/maven-plugin/build.gradle.kts]: Use import.

File: build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    val dokka_version: String by this
+
+    if (this == rootProject) {
+        println("Publication version: $dokka_version")
+    }
+    group = "org.jetbrains.dokka"
+    version = dokka_version
+
+    val language_version: String by project
+    tasks.withType(org.jetbrains.kotlin.gradle.tasks.KotlinCompile::class).all {
```
[the remark this answers is not in the document]
Comment (kamilok1965) [reply build.gradle.kts]: Fixed

File: build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    id("com.jfrog.bintray") apply false
+}
+
+allprojects {
+    configureDokkaVersion()
+    val dokka_version: String by this
+
+    if (this == rootProject) {
+        println("Publication version: $dokka_version")
+    }
```
[the remark this answers is not in the document]
Comment (kamilok1965) [reply build.gradle.kts]: I wanted to retain some order with configuring dokka version but I agree, it'll be cleaner this way

File: buildSrc/src/main/kotlin/org/jetbrains/BintrayPublishing.kt [hunk 4]
```diff
@@ -0,0 +1,29 @@
+package org.jetbrains
+
+import com.jfrog.bintray.gradle.BintrayExtension
+import org.gradle.api.Project
+import org.gradle.kotlin.dsl.provideDelegate
+
+fun Project.configureBintrayPublication(publication: String) = configureBintrayPublication(listOf(publication))
+
+fun Project.configureBintrayPublication(publications: List<String>) {
```
[the remark this answers is not in the document]
Comment (kamilok1965) [reply buildSrc/src/main/kotlin/org/jetbrains/BintrayPublishing.kt]: Done

File: buildSrc/src/main/kotlin/org/jetbrains/DistMavenPublishing.kt [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+import org.gradle.api.artifacts.repositories.MavenArtifactRepository
+import org.gradle.api.publish.PublishingExtension
+import org.gradle.api.publish.maven.tasks.PublishToMavenRepository
+
+fun Project.configureDistMaven() { // TODO: This can probably be written cleaner
+    val repoLocation = uri(file("${rootProject.buildDir}/dist-maven"))
+    var distMaven: MavenArtifactRepository? = null
+    pluginManager.withPlugin("maven-publish") {
+        this@configureDistMaven.extensions.findByType(PublishingExtension::class.java)?.let {
+            it.repositories {
```
[the remark this answers is not in the document]
Comment (kamilok1965) [reply buildSrc/src/main/kotlin/org/jetbrains/DistMavenPublishing.kt]: That's true, fixed

File: gradle.properties [hunk 6]
```diff
@@ -2,22 +2,19 @@ dokka_version_base=0.11.0
...
 
 # Maven plugin dependencies
 maven_version=3.5.0
 maven_archiver_version=2.5
-plexus_utils_version=3.0.22
-plexus_archiver_version=3.4
 maven_plugin_tools_version=3.5.2
 
 # For CI
 mvn=mvn
```
[the remark this answers is not in the document]
Comment (kamilok1965) [reply gradle.properties]: Sure thing

[the remark this answers is not in the document]
Comment (kamilok1965) [reply gradle.properties]: Should I inline `maven_version` and `plugin_tools` in maven-plugin's build.gradle.kts?

File: integration-tests/gradle-integration-tests/build.gradle.kts [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+    kotlinGradle("org.jetbrains.kotlin:kotlin-gradle-plugin")
+
+    testImplementation(group = "junit", name = "junit", version = "4.13")
+    testImplementation(gradleTestKit())
+}
+
+
+val createClasspathManifest by tasks.registering {
+    dependsOn(project(":core").getTasksByName("shadowJar", true))
```
[the remark this answers is not in the document]
Comment (kamilok1965) [reply integration-tests/gradle-integration-tests/build.gradle.kts]: Yes, they should be regenerated automatically. Removed

File: runners/maven-plugin/build.gradle.kts (hunk 1 above)
Comment (kamilok1965) [reply runners/maven-plugin/build.gradle.kts]: Done

File: gradle.properties (hunk 6 above)
[the remark this answers is not in the document]
Comment (Kordyjan) [reply gradle.properties]: I think it is good idea.

File: integration-tests/gradle-integration-tests/build.gradle.kts [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+        outputDir.mkdirs()
+        file("$outputDir/dokka-plugin-classpath.txt").writeText(dokkaPlugin.joinToString("\n"))
+        file("$outputDir/fatjar.txt").writeText(dokkaCore.joinToString("\n"))
+        file("$outputDir/kotlin-gradle.txt").writeText(kotlinGradle.joinToString("\n"))
+    }
+}
+
+val testClasses by tasks.getting
+
+testClasses.dependsOn(project(":core").getTasksByName("shadowJar", true))
```
[the remark this answers is not in the document]
Comment (kamilok1965) [reply integration-tests/gradle-integration-tests/build.gradle.kts]: I don't know TBH, this file was just migrated as it was. Looks like we can even omit entirely and just have it in the classpath manifest for now as the tests use the core only in the runtime

File: gradle.properties (hunk 6 above)
[the remark this answers is not in the document]
Comment (kamilok1965) [reply gradle.properties]: Done

File: buildSrc/build.gradle.kts [hunk 9]
```diff
@@ -0,0 +1,14 @@
+plugins {
+    `kotlin-dsl`
+    groovy
```
[the remark this answers is not in the document]
Comment (kamilok1965) [reply buildSrc/build.gradle.kts]: Yup

**Pull request merged**